### PR TITLE
[maxtext] Add block-diffusion training primitives

### PR DESCRIPTION
**Pull request by ethannnnnn**
## Motivation

MaxText needs small, model-independent primitives for discrete block-diffusion
training before attention, input-pipeline, trainer, or post-training support is
introduced.

## Scope

This PR adds only:

- block corruption under `maxtext.diffusion.block_diffusion`;
- explicit target alignment for `same_position` and `shifted` logits; and
- focused unit tests for both supported contracts.

It does not change configuration, attention, model execution, training loops,
serving, SFT, distillation, or RL.

## Design

The corruption API consumes clean tokens, an explicit eligibility mask, a
logical block size, and one of two declared model contracts:

- `same_position` with an `all_masked` canvas; or
- `shifted` with a `seed_and_mask` canvas.

Each nonempty eligible block is guaranteed to retain at least one supervised
position. Padding remains untouched. Target alignment follows logical token
positions, including reordered physical layouts, instead of assuming that
array index equals sequence position.

## Compatibility

There is no behavior change for existing MaxText training or inference paths.
The new package is not activated by any configuration in this PR.

## Tests

`python -m pytest -q tests/unit/block_diffusion_test.py`

The tests cover partial blocks, per-block noise sampling, deterministic RNG,
all-invalid rows, seeded anchors, mask/loss provenance, reordered logical
positions, padding, and invalid contracts and shapes.

Current clean-head result: `13 passed`, `5 subtests passed`; Pyink, compileall,
and `git diff --check` also pass.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added comments or docstrings where the contracts are not self-evident.
- [x] End-to-end testing is not applicable to this default-off, pure primitive layer; focused CPU tests are listed above.
- [x] No standalone documentation page or toctree change is needed; the public design document is linked below.

## Follow-ups

Block-causal attention and pre-training integration will be proposed only after
this primitive contract is reviewed and landed. CFT/SFT, OPD, exact replay,
and RL remain separate later changes.

Design document:
https://docs.google.com/document/d/1N7KcCoAIErB2CV9EJ2G1u_mdN-AQgqwNUYSvM0mtqMI/edit

### CODE REVIEW

File: src/maxtext/diffusion/block_diffusion/corruption.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    rng: np.random.Generator,
+    *,
+    block_size: int,
+    mask_id: int,
+    min_noise: float = 1.0e-3,
+    logit_alignment: str = "same_position",
+    canvas_policy: str = "all_masked",
+    axis: int = 1,
+) -> BlockDiffusionCorruptionResult:
+  """Masks valid tokens independently within each logical diffusion block."""
```
Comment (hengtaoguo) [inline src/maxtext/diffusion/block_diffusion/corruption.py]: Could you add a few doc string to preview the expected shape of `inputs/validity_mask`? Something like [B, S, ...]. Same for the `align_logits_to_targets() - logits` below. Thanks!

**Review by hengtaoguo — APPROVED** [review]
Thanks for the feature contribution!

File: tests/unit/block_diffusion_test.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    del low, high
+    self.uniform_calls += 1
+    return np.asarray([[[0.1], [0.9], [0.1]], [[0.1], [0.9], [0.1]]]).reshape(size)
+
+  def random(self, size):
+    self.random_calls += 1
+    return np.full(size, 0.5)
+
+
+class BlockDiffusionCorruptionTest(unittest.TestCase):
```
Comment (hengtaoguo) [inline tests/unit/block_diffusion_test.py:68]: Could we mark this test `cpu_only` just to save the CI resources? Unless there's a strong preference to have it on TPU. WDYT?

cc @entrpn

File: src/maxtext/diffusion/block_diffusion/corruption.py (hunk 1 above)
Comment (ethannnnnn) [reply src/maxtext/diffusion/block_diffusion/corruption.py]: Addressed in the latest update. The docstrings now document the `[batch, sequence, ...]` token/mask contract and the `[batch, sequence, vocabulary]` logits contract.

File: tests/unit/block_diffusion_test.py (hunk 2 above)
Comment (ethannnnnn) [reply tests/unit/block_diffusion_test.py:68]: Done. The test module now uses `pytest.mark.cpu_only`, so it is skipped on TPU and GPU testbeds.

File: src/maxtext/diffusion/block_diffusion/target_alignment.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Target alignment for discrete block-diffusion objectives."""
+
+import jax.numpy as jnp
+
+
+def align_logits_to_targets(logits, alignment, positions=None, validity_mask=None):
```
Comment (shralex) [inline src/maxtext/diffusion/block_diffusion/target_alignment.py]: could you please add type hints like you did in other functions in this pr

**Review by shralex — COMMENTED** [review]
Since block diffusion deviates significantly from the standard autoregressive (AR) pipeline (introducing concepts like canvas_policy, seed_and_mask, and target_alignment), future contributors would greatly benefit from a README.md in the src/maxtext/diffusion/block_diffusion/ directory.

You could include the following in the README or module-level docstring:

Literature Links: Link to the Block Diffusion (https://arxiv.org/abs/2503.09573 ?) paper so readers understand the theoretical backing.

Terminology definitions: Briefly explain the core concepts implemented in this PR:
Canvas: The fixed-length sequence of token slots being refined via denoising.
Logit Alignment (same_position vs shifted): Explain how logits map to targets, as standard causal AR models shift targets by 1, whereas diffusion models might align them directly.
Corruption / Masking: Explain that corrupt_tokens handles the forward noising process by masking valid tokens independently within each bounded bidirectional block.

Adding this context will make the codebase much more approachable for anyone maintaining the diffusion or speculative decoding pipelines in MaxText!

File: src/maxtext/diffusion/block_diffusion/target_alignment.py (hunk 3 above)
Comment (ethannnnnn) [reply src/maxtext/diffusion/block_diffusion/target_alignment.py]: Done. `align_logits_to_targets` now annotates the JAX arrays, alignment string, optional position and validity arrays, and return type.

**Comment by ethannnnnn** [comment]
Thanks for the documentation suggestion. I added `src/maxtext/diffusion/block_diffusion/README.md` with the block-diffusion paper, canvas and block terminology, corruption semantics, logit alignment, and the two currently supported training contracts.

**Review by shralex — APPROVED** [review]

**Comment by ethannnnnn** [comment]
@hengtaoguo @shralex Thanks for the approvals! All required checks are green and the PR is pull-ready. GitHub prevents me from merging into protected main. Could one of you please squash-merge it or enable auto-merge?
